src/stm32f411/dma_reg.hpp

```diff
@@ -0,0 +1,148 @@
+  /// 0: DMA is the flow controller
+  /// 1: peripheral is the flow controller
+  static constexpr auto peripheral_flow_controller = bit_mask::from<5>();
+  /// 00: Peripheral-to-memory
+  /// 01: Memory-to-peripheral
+  /// 10: Memory-to-memory
+  static constexpr auto data_transfer_direction = bit_mask::from<5>();
+  static constexpr auto circular_mode_enable = bit_mask::from<8>();
+  /// 0: peripheral address pointer is fixed
+  /// 1: peripheral address pointer is incremented after each data transfer (set
+  /// by peripheral_data_size)
+  static constexpr auto peripheral_increment_mode = bit_mask::from<9>();
+  /// 0: memory address pointer is fixed
+  /// 1: memory address pointer is incremented after each data transfer (set by
+  /// memory_data_size)
+  static constexpr auto memory_increment_mode = bit_mask::from<10>();
+  /// 00: byte (8-bit)
+  /// 01: half-word (16-bit)
+  /// 10: word (32-bit)
+  static constexpr auto peripheral_data_size = bit_mask::from<12, 11>();
+  /// 00: byte (8-bit)
+  /// 01: half-word (16-bit)
+  /// 10: word (32-bit)
+  static constexpr auto memory_data_size = bit_mask::from<14, 11>();
+  /// 0: The offset size for the peripheral address calculation is linked to the
+  /// PSIZE 1: The offset size for the peripheral address calculation is fixed
+  /// to 4 (32-bit alignment).
+  static constexpr auto peripheral_offset_size = bit_mask::from<15>();
+  /// 00: Low
+  /// 01: Medium
+  /// 10: High
+  /// 11: Very high
+  static constexpr auto priority_level = bit_mask::from<17, 16>();
+  /// 0: No buffer switching at the end of transfer
+  /// 1: Memory target switched at the end of the DMA transfer
+  static constexpr auto double_buffer_mode = bit_mask::from<18>();
+  /// 0: The current target memory is Memory 0 (addressed by the DMA_SxM0AR
+  /// pointer) 1: The current target memory is Memory 1 (addressed by the
+  /// DMA_SxM1AR pointer)
+  static constexpr auto current_target = bit_mask::from<19>();
+  /// 00: single transfer
+  /// 01: INCR4 (incremental burst of 4 beats)
+  /// 10: INCR8 (incremental burst of 8 beats)
+  /// 11: INCR16 (incremental burst of 16 beats)
+  static constexpr auto peripheral_burst_transfer = bit_mask::from<22, 21>();
+  /// 00: single transfer
+  /// 01: INCR4 (incremental burst of 4 beats)
+  /// 10: INCR8 (incremental burst of 8 beats)
+  /// 11: INCR16 (incremental burst of 16 beats)
+  static constexpr auto memory_burst_transfer = bit_mask::from<24, 23>();
+  /// Refer to Table 27(DMA 1) or 28(DMA 2) of the user manual for channel
+  /// selection
+  static constexpr auto channel_select = bit_mask::from<17, 25>();
+};
+
+inline constexpr intptr_t ahb_base = 0x4002'0000UL;
+static inline dma_config_t* get_dma_reg(peripheral p_dma)
+{
+  // STM has dedicated memory blocks where every 2^10 is a new DMA register
+  return reinterpret_cast<dma_config_t*>(
+    ahb_base + 0x6000 +
+    ((static_cast<int>(p_dma) - static_cast<int>(peripheral::dma1)) << 10));
+}
```

Comment:
Just a style thing. Could you break this up into a sequence of named `constexpr` variables to make the math a bit easier to read.